message_manager.go

```diff
@@ -0,0 +1,324 @@
+package dht
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/libp2p/go-libp2p-kad-dht/metrics"
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+
+	"github.com/libp2p/go-msgio"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+type messageManager struct {
+	host      host.Host // the network services we need
+	smlk      sync.Mutex
+	strmap    map[peer.ID]*messageSender
+	protocols []protocol.ID
+}
+
+func (m *messageManager) streamDisconnect(ctx context.Context, p peer.ID) {
+	m.smlk.Lock()
+	defer m.smlk.Unlock()
+	ms, ok := m.strmap[p]
+	if !ok {
+		return
+	}
+	delete(m.strmap, p)
+
+	// Do this asynchronously as ms.lk can block for a while.
+	go func() {
+		if err := ms.lk.Lock(ctx); err != nil {
+			return
+		}
+		defer ms.lk.Unlock()
+		ms.invalidate()
+	}()
+}
+
+// SendRequest sends out a request, but also makes sure to
```

Comment:
this is a somewhat weird place to be doing this. It
* isn't DHT specific - why aren't we doing this at the host connection level?
* feels potentially ugly with a general connection manager rather than factored out into a peer prioritization logic place.

I think the Ask is to think about what the ideal place for speed measurement logic to live, and file an issue for that (probably not in this repo.) I'm okay having it here for now

Reply:
This was just a move from `dht_net.go` where it was previously. I tend to agree that I'm not sure why we're recording latency with `m.host.Peerstore().RecordLatency(p, time.Since(start))` when it includes processing time on the server too.

The other metrics emitted here are DHT specific though.